src/main/java/net/minecraftforge/datafix/ForgeSchema.java

```diff
@@ -0,0 +1,298 @@
+package net.minecraftforge.datafix;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.mojang.datafixers.DSL;
+import com.mojang.datafixers.DataFixUtils;
+import com.mojang.datafixers.schemas.Schema;
+import com.mojang.datafixers.types.Type;
+import com.mojang.datafixers.types.families.RecursiveTypeFamily;
+import com.mojang.datafixers.types.families.TypeFamily;
+import com.mojang.datafixers.types.templates.RecursivePoint;
+import com.mojang.datafixers.types.templates.TaggedChoice;
+import com.mojang.datafixers.types.templates.TypeTemplate;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * A partial carbon copy of the {@link Schema} class.
+ * But this allows us to delay the setup of the type hierarchy until we need it.
+ */
+class ForgeSchema extends Schema
+{
+    //Can not be final, needs to be initialized on the fly!
+    //Constructor invocation order and the what not!
+    private Map<String, Supplier<TypeTemplate>> TYPE_TEMPLATES = Maps.newConcurrentMap();
+    private Map<String, Type<?>>                TYPES          = Maps.newConcurrentMap();
+    private Map<String, Integer>                RECURSIVE_TYPES = Maps.newConcurrentMap();
+
+    private final Map<String, Supplier<TypeTemplate>> MODDED_ENTITY_TYPES = Maps.newConcurrentMap();
+    private final Map<String, Supplier<TypeTemplate>> MODDED_BLOCK_ENTITY_TYPES = Maps.newConcurrentMap();
+
+    //Store some data we can not access later anymore.
+    private final String name;
+    private final Schema wrapped;
+
+    ForgeSchema(final int versionKey, final Schema parent, final Schema wrapped)
+    {
+        super(versionKey, parent);
+        this.wrapped = wrapped;
+
+        final int subVersion = DataFixUtils.getSubVersion(versionKey);
+        name = "V" + DataFixUtils.getVersion(versionKey) + (subVersion == 0 ? "" : "." + subVersion);
+
+        //Now with the vanilla data in the schema, lets build our first version again.
+        //Yes this executes the type resolve multiple times per schema (but it's not a big deal)
+        resetSchema();
+        rebuildSchema(
+          Collections.emptyMap(),
+          Collections.emptyMap()
+        );
+    }
+
+    /**
+     * Rebuilds the types, with support for the case that no types exists.
+     * (Which should never happen, but you never know!)
+     *
+     * @return The type map.
+     */
+    protected Map<String, Type<?>> buildTypes()
+    {
+        //Result map.
+        final Map<String, Type<?>> types = Maps.newHashMap();
```

Comment:
Is there a reason that this method returns a hashmap, but the variable that is set to this is defaulted to a concurrent map?

Reply:
Nope just missed it.